--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -720,17 +720,6 @@ static int spidev_probe(struct spi_devic
 	int			status;
 	unsigned long		minor;
 
-	/*
-	 * spidev should never be referenced in DT without a specific
-	 * compatible string, it is a Linux implementation thing
-	 * rather than a description of the hardware.
-	 */
-	if (spi->dev.of_node && !of_match_device(spidev_dt_ids, &spi->dev)) {
-		dev_err(&spi->dev, "buggy DT: spidev listed directly in DT\n");
-		WARN_ON(spi->dev.of_node &&
-			!of_match_device(spidev_dt_ids, &spi->dev));
-	}
-
 	spidev_probe_acpi(spi);
 
 	/* Allocate driver data */
