--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -720,15 +720,6 @@ static int spidev_probe(struct spi_devic
 	int			status;
 	unsigned long		minor;
 
-	/*
-	 * spidev should never be referenced in DT without a specific
-	 * compatible string, it is a Linux implementation thing
-	 * rather than a description of the hardware.
-	 */
-	WARN(spi->dev.of_node &&
-	     of_device_is_compatible(spi->dev.of_node, "spidev"),
-	     "%pOF: buggy DT: spidev listed directly in DT\n", spi->dev.of_node);
-
 	spidev_probe_acpi(spi);
 
 	/* Allocate driver data */
