--- a/ap.c
+++ b/ap.c
@@ -7636,6 +7636,8 @@ skip_key_mgmt:
 
 	if (dut->program == PROGRAM_VHT) {
 		int vht_oper_centr_freq_idx;
+		char vht_capab[256] = "";
+		int len = 0;
 
 		if (check_channel(dut->ap_channel) < 0) {
 			send_resp(dut, conn, SIGMA_INVALID,
@@ -7682,15 +7684,19 @@ skip_key_mgmt:
 
 		if (dut->ap_sgi80 || dut->ap_txBF ||
 		    dut->ap_ldpc != VALUE_NOT_SET ||
-		    dut->ap_tx_stbc || dut->ap_mu_txBF) {
-			fprintf(f, "vht_capab=%s%s%s%s%s\n",
-				dut->ap_sgi80 ? "[SHORT-GI-80]" : "",
-				dut->ap_txBF ?
-				"[SU-BEAMFORMER][SU-BEAMFORMEE][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
-				(dut->ap_ldpc == VALUE_ENABLED) ?
-				"[RXLDPC]" : "",
-				dut->ap_tx_stbc ? "[TX-STBC-2BY1][RX-STBC-1]" : "",
-				dut->ap_mu_txBF ? "[MU-BEAMFORMER]" : "");
+		    dut->ap_tx_stbc || dut->ap_mu_txBF ||
+		    dut->ap_ampdu_exp) {
+			len += snprintf(vht_capab + len, sizeof(vht_capab) - len, "%s%s%s%s%s",
+					dut->ap_sgi80 ? "[SHORT-GI-80]" : "",
+					dut->ap_txBF ?
+					"[SU-BEAMFORMER][SU-BEAMFORMEE][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
+					(dut->ap_ldpc == VALUE_ENABLED) ?
+					"[RXLDPC]" : "",
+					dut->ap_tx_stbc ? "[TX-STBC-2BY1][RX-STBC-1]" : "",
+					dut->ap_mu_txBF ? "[MU-BEAMFORMER]" : "");
+			len += snprintf(vht_capab + len, sizeof(vht_capab) - len,
+					"[MAX-A-MPDU-LEN-EXP%d][MAX-MPDU-11454]", dut->ap_ampdu_exp);
+			fprintf(f, "vht_capab=%s\n", vht_capab);
 		}
 	}
 
@@ -8310,6 +8316,7 @@ static enum sigma_cmd_result cmd_ap_rese
 	dut->ap_txBF = 0;
 	dut->ap_mu_txBF = 0;
 	dut->ap_chwidth = AP_AUTO;
+	dut->ap_ampdu_exp = 7;
 
 	dut->ap_rsn_preauth = 0;
 	dut->ap_wpsnfc = 0;
