--- a/ap.c
+++ b/ap.c
@@ -9088,6 +9088,10 @@ skip_key_mgmt:
 						   ifname, dut->he_mcs);
 		}
 
+		if (dut->he_sounding == VALUE_ENABLED)
+			run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 7 0",
+					   ifname);
+
 	}
 
 	if (dut->program == PROGRAM_60GHZ) {
@@ -9663,6 +9667,7 @@ static enum sigma_cmd_result cmd_ap_rese
 	}
 
 	if (dut->program == PROGRAM_HE) {
+		dut->he_ackpolicymac = 0;
 		if (dut->device_type == AP_testbed) {
 			dut->ap_ldpc = VALUE_DISABLED;
 			dut->ap_ba_bufsize = BA_BUFSIZE_64;
@@ -13885,6 +13890,40 @@ static int mac80211_set_trigger_type(str
 	return SUCCESS_SEND_STATUS;
 }
 
+static void mac80211_set_ackpolicy_0(struct sigma_dut *dut, const char *ifname,
+				     unsigned char *mac_addr)
+{
+	if (dut->he_ackpolicymac != 1)
+		return;
+
+	/*Disable all-BAR ackpolicy for MUMIMO*/
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 62 0",
+			   ifname);
+	/*Disable all-BAR ackpolicy first*/
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 64 0",
+			   ifname);
+	/*Set normal ack policy for the STA with the specified mac address in dl-tx case*/
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x4b -v 0 8 1 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x 0x%02x",
+			   ifname, mac_addr[0], mac_addr[1], mac_addr[2],
+			   mac_addr[3], mac_addr[4], mac_addr[5]);
+}
+
+static void mac80211_set_ackpolicy_3(struct sigma_dut *dut, const char *ifname)
+{
+	/*Enable all-BAR ackpolicy for MUMIMO DL*/
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 62 1",
+			   ifname);
+	/*Enable all-BAR ackpolicy*/
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x48 -v 0 64 1",
+			   ifname);
+}
+
+static void mac80211_set_ackpolicy_4(struct sigma_dut *dut, const char *ifname)
+{
+	/*Enable htp-ack ackpolicy*/
+	run_system_wrapper(dut, "ath11k-fwtest -i %s -m 0x47 -v 0 99 1",
+			   ifname);
+}
 
 static int16_t mac80211_he_ltf_mapping(struct sigma_dut *dut,
 				       struct sigma_conn *conn,
@@ -13967,6 +14006,7 @@ static int mac80211_ap_set_rfeature(stru
 {
 	const char *val, *ifname;;
 	char *type = "he";
+	unsigned char mac_addr[ETH_ALEN];
 
 	ifname = get_main_ifname(dut);
 
@@ -14031,6 +14071,37 @@ static int mac80211_ap_set_rfeature(stru
 		return STATUS_SENT_ERROR;
 	}
 
+	val = get_param(cmd, "AckPolicy_MAC");
+	if (val) {
+		dut->he_ackpolicymac = 1;
+		memset(mac_addr, '0', ETH_ALEN);
+		if (parse_mac_address(dut, val, mac_addr) < 0) {
+			sigma_dut_print(dut, DUT_MSG_ERROR,
+					"MAC Address not in proper format");
+			return INVALID_SEND_STATUS;
+		}
+	}
+
+	val = get_param(cmd, "AckPolicy");
+	if (val) {
+		dut->ap_he_ackpolicy = atoi(val);
+
+		switch (dut->ap_he_ackpolicy) {
+		case 0:
+			mac80211_set_ackpolicy_0(dut, ifname, mac_addr);
+			break;
+		case 3:
+			mac80211_set_ackpolicy_3(dut, ifname);
+			break;
+		case 4:
+			mac80211_set_ackpolicy_4(dut, ifname);
+			break;
+		default:
+			sigma_dut_print(dut, DUT_MSG_ERROR,
+					"Invalid AckPolicy setting");
+		}
+	}
+
 	/* Extract LTF value before GI because in case of auto-rate,
 	 * dut->ar_ltf should be set before calling
 	 * mac80211_he_ar_gi_ltf_mask()
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -938,6 +938,8 @@ struct sigma_dut {
 	bool wifi_hal_initialized;
 #endif /*ANDROID_WIFI_HAL*/
 	int he_mcs;
+	int he_ackpolicymac;
+	int ap_he_ackpolicy;
 	int sae_h2e_default;
 	enum {
 		SAE_PWE_DEFAULT,
