diff '--color=auto' -aur a/boot/image.c b/boot/image.c
--- a/boot/image.c	2022-09-07 05:52:05.468537304 -0400
+++ b/boot/image.c	2022-09-07 05:50:16.084542072 -0400
@@ -91,6 +91,7 @@
 	{	IH_ARCH_OPENRISC,	"or1k",		"OpenRISC 1000",},
 	{	IH_ARCH_SANDBOX,	"sandbox",	"Sandbox",	},
 	{	IH_ARCH_ARM64,		"arm64",	"AArch64",	},
+	{	IH_ARCH_ARM64,		"aarch64",	"AArch64",	},
 	{	IH_ARCH_ARC,		"arc",		"ARC",		},
 	{	IH_ARCH_X86_64,		"x86_64",	"AMD x86_64",	},
 	{	IH_ARCH_XTENSA,		"xtensa",	"Xtensa",	},
