diff '--color=auto' -aur a/session.c b/session.c
--- a/session.c	2020-05-26 11:24:24.000000000 -0400
+++ b/session.c	2022-09-06 22:07:42.397504073 -0400
@@ -226,7 +226,7 @@
 
 	blobmsg_add_string(&buf, "ubus_rpc_session", ses->id);
 	blobmsg_add_u32(&buf, "timeout", ses->timeout);
-	blobmsg_add_u32(&buf, "expires", uloop_timeout_remaining(&ses->t) / 1000);
+	blobmsg_add_u32(&buf, "expires", uloop_timeout_remaining64(&ses->t) / 1000);
 
 	if (acls) {
 		c = blobmsg_open_table(&buf, "acls");
