--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -831,6 +831,15 @@ fail:
 	 * hostapd_setup_interface_complete() will return -1 on failure,
 	 * 0 on success and 0 is HOSTAPD_CHAN_VALID :)
 	 */
+
+	if (err != 0 && (iface->conf->secondary_channel || iface->conf->ieee80211ac)) {
+		wpa_printf(MSG_ERROR, "ACS: Fallback to 20MHz channel bandwidth, will retry interface setup");
+		iface->conf->secondary_channel = 0;
+		iface->conf->ieee80211ac = 0;
+		hostapd_select_hw_mode(iface);
+		return;
+	}
+
 	if (hostapd_acs_completed(iface, err) == HOSTAPD_CHAN_VALID) {
 		acs_cleanup(iface);
 		return;
