diff -urN a/drivers/mtd/devices/Kconfig b/drivers/mtd/devices/Kconfig
--- a/drivers/mtd/devices/Kconfig	2016-02-24 19:59:14.152108305 +0800
+++ b/drivers/mtd/devices/Kconfig	2016-02-24 19:59:01.356108095 +0800
@@ -80,7 +80,7 @@

 config MTD_M25P80
	tristate "Support most SPI Flash chips (AT26DF, M25P, W25X, ...)"
-	depends on SPI_MASTER && MTD_SPI_NOR
+	depends on SPI_MASTER
	help
	  This enables access to most modern SPI flash chips, used for
	  program and data storage.   Series supported include Atmel AT26DF,
@@ -95,6 +95,55 @@
	  if you want to specify device partitioning or to use a device which
	  doesn't support the JEDEC ID instruction.

+config M25PXX_USE_USER_MODE
+	bool "Support SPI User Mode for sheipa"
+	depends on MTD_M25P80
+	default y
+	help
+	  This option enables user mode to do write/read operation by ST M25Pxx.
+
+config M25PXX_USE_FAST_READ
+	bool "Use FAST_READ OPCode allowing SPI CLK >= 50MHz"
+	depends on MTD_M25P80
+	depends on M25PXX_USE_USER_MODE
+	default n
+	help
+	  This option enables FAST_READ access supported by ST M25Pxx.
+
+config M25PXX_USE_MULTI_CHANNEL
+	bool "Support Multi-channel write/read"
+	depends on MTD_M25P80
+	depends on M25PXX_USE_USER_MODE
+	default n
+	help
+	  This option enables Multi-channel write/read access supported by ST M25Pxx.
+
+config M25PXX_USE_QUAD
+	bool "Use QUAD OPCode allowing SPI do quad-channel write/read"
+	depends on !RTL_8197F
+	depends on MTD_M25P80
+	depends on M25PXX_USE_USER_MODE
+	depends on M25PXX_USE_MULTI_CHANNEL
+	default n
+	help
+	  This option enables quad-channel access supported by ST M25Pxx.
+
+config M25PXX_USE_DUAL
+	bool "Use DUAL OPCode allowing SPI do dual-channel write/read"
+	depends on MTD_M25P80
+	depends on M25PXX_USE_USER_MODE
+	depends on M25PXX_USE_MULTI_CHANNEL
+	default n
+	help
+	  This option enables dual-channel access supported by ST M25Pxx.
+
+config M25PXX_USE_AUTO_MODE
+	bool "Support SPI Auto Mode for sheipa"
+	depends on MTD_M25P80
+	default n
+	help
+	  This option enables auto mode to do auto write/read operation by ST M25Pxx.
+
 config MTD_SPEAR_SMI
	tristate "SPEAR MTD NOR Support through SMI controller"
	depends on PLAT_SPEAR
@@ -102,6 +151,14 @@
	help
	  This enable SNOR support on SPEAR platforms using SMI controller

+config M25PXX_PREFER_SMALL_SECTOR_ERASE
+	bool "Prefer small sector erase"
+	depends on MTD_M25P80
+	default y
+	help
+	  This option enables use of the small erase sectors if that is
+	  supported by the flash chip.
+
 config MTD_SST25L
	tristate "Support SST25L (non JEDEC) SPI Flash chips"
	depends on SPI_MASTER
@@ -210,14 +267,6 @@
	  M-Systems and now Sandisk. The support is very experimental,
	  and doesn't give access to any write operations.

-config MTD_ST_SPI_FSM
-	tristate "ST Microelectronics SPI FSM Serial Flash Controller"
-	depends on ARCH_STI
-	help
-	  This provides an MTD device driver for the ST Microelectronics
-	  SPI Fast Sequence Mode (FSM) Serial Flash Controller and support
-	  for a subset of connected Serial Flash devices.
-
 if MTD_DOCG3
 config BCH_CONST_M
	default 14
@@ -225,4 +274,59 @@
	default 4
 endif

+config MTD_DOCPROBE
+	tristate
+	select MTD_DOCECC
+
+config MTD_DOCECC
+	tristate
+
+config MTD_DOCPROBE_ADVANCED
+	bool "Advanced detection options for DiskOnChip"
+	depends on MTD_DOCPROBE
+	help
+	  This option allows you to specify nonstandard address at which to
+	  probe for a DiskOnChip, or to change the detection options.  You
+	  are unlikely to need any of this unless you are using LinuxBIOS.
+	  Say 'N'.
+
+config MTD_DOCPROBE_ADDRESS
+	hex "Physical address of DiskOnChip" if MTD_DOCPROBE_ADVANCED
+	depends on MTD_DOCPROBE
+	default "0x0"
+	---help---
+	  By default, the probe for DiskOnChip devices will look for a
+	  DiskOnChip at every multiple of 0x2000 between 0xC8000 and 0xEE000.
+	  This option allows you to specify a single address at which to probe
+	  for the device, which is useful if you have other devices in that
+	  range which get upset when they are probed.
+
+	  (Note that on PowerPC, the normal probe will only check at
+	  0xE4000000.)
+
+	  Normally, you should leave this set to zero, to allow the probe at
+	  the normal addresses.
+
+config MTD_DOCPROBE_HIGH
+	bool "Probe high addresses"
+	depends on MTD_DOCPROBE_ADVANCED
+	help
+	  By default, the probe for DiskOnChip devices will look for a
+	  DiskOnChip at every multiple of 0x2000 between 0xC8000 and 0xEE000.
+	  This option changes to make it probe between 0xFFFC8000 and
+	  0xFFFEE000.  Unless you are using LinuxBIOS, this is unlikely to be
+	  useful to you.  Say 'N'.
+
+config MTD_DOCPROBE_55AA
+	bool "Probe for 0x55 0xAA BIOS Extension Signature"
+	depends on MTD_DOCPROBE_ADVANCED
+	help
+	  Check for the 0x55 0xAA signature of a DiskOnChip, and do not
+	  continue with probing if it is absent.  The signature will always be
+	  present for a DiskOnChip 2000 or a normal DiskOnChip Millennium.
+	  Only if you have overwritten the first block of a DiskOnChip
+	  Millennium will it be absent.  Enable this option if you are using
+	  LinuxBIOS or if you need to recover a DiskOnChip Millennium on which
+	  you have managed to wipe the first block.
+
 endmenu
diff -urN a/drivers/mtd/devices/m25p80.c b/drivers/mtd/devices/m25p80.c
--- a/drivers/mtd/devices/m25p80.c	2016-02-24 19:59:29.936108565 +0800
+++ b/drivers/mtd/devices/m25p80.c	2016-02-24 19:58:51.168107928 +0800
@@ -15,102 +15,862 @@
  *
  */

+#include <linux/init.h>
 #include <linux/err.h>
 #include <linux/errno.h>
 #include <linux/module.h>
 #include <linux/device.h>
+#include <linux/interrupt.h>
+#include <linux/mutex.h>
+#include <linux/math64.h>
+#include <linux/slab.h>
+#include <linux/sched.h>
+#include <linux/mod_devicetable.h>

+#include <linux/mtd/cfi.h>
 #include <linux/mtd/mtd.h>
 #include <linux/mtd/partitions.h>
+#include <linux/of_platform.h>

 #include <linux/spi/spi.h>
 #include <linux/spi/flash.h>
-#include <linux/mtd/spi-nor.h>

+/* Flash opcodes. */
+#define	OPCODE_WREN		0x06	/* Write enable */
+#define	OPCODE_RDSR		0x05	/* Read status register */
+#define	OPCODE_WRSR		0x01	/* Write status register 1 byte */
+#define	OPCODE_NORM_READ	0x03	/* Read data bytes (low frequency) */
+#define	OPCODE_FAST_READ	0x0b	/* Read data bytes (high frequency) */
+#define	OPCODE_QUAD_READ        0x6b    /* Read data bytes */
+#define	OPCODE_PP		0x02	/* Page program (up to 256 bytes) */
+#define	OPCODE_BE_4K		0x20	/* Erase 4KiB block */
+#define	OPCODE_BE_4K_PMC	0xd7	/* Erase 4KiB block on PMC chips */
+#define	OPCODE_BE_32K		0x52	/* Erase 32KiB block */
+#define	OPCODE_CHIP_ERASE	0xc7	/* Erase whole flash chip */
+#define	OPCODE_SE		0xd8	/* Sector erase (usually 64KiB) */
+#define	OPCODE_RDID		0x9f	/* Read JEDEC ID */
+#define	OPCODE_RDCR             0x35    /* Read configuration register */
+
+/* 4-byte address opcodes - used on Spansion and some Macronix flashes. */
+#define	OPCODE_NORM_READ_4B	0x13	/* Read data bytes (low frequency) */
+#define	OPCODE_FAST_READ_4B	0x0c	/* Read data bytes (high frequency) */
+#define	OPCODE_QUAD_READ_4B	0x6c    /* Read data bytes */
+#define	OPCODE_PP_4B		0x12	/* Page program (up to 256 bytes) */
+#define	OPCODE_SE_4B		0xdc	/* Sector erase (usually 64KiB) */
+
+/* Support auto mode for sheipa only. */
+#define	OPCODE_AUTO		0xf5
+
+/* Extend flash opcode and used for Macronix */
+#define PPX2_I		0x02		/* 1|2O dual program */
+#define PPX2_II		0x02		/* 2 x I/O dual program */
+#define PPX4_I		0x02		/* 1|4O quad program */
+#define PPX4_II		0x38		/* 4 x I/O quad program */
+#define READX2_I	0x03		/* 1|2O dual dual read*/
+#define READX2_IO	0xbb		/* 2 x I/O dual read */
+#define READX4_I	0x03		/* 1|4O quad read */
+#define READX4_IO	0xeb		/* 4 x I/O quad read */
+
+/* Extend flash read dummy cycles and used for Macronix only */
+#define MXIC_DUAL_DUMMY_CYCLE    0x4
+#define MXIC_QUAD_DUMMY_CYCLE    0x6
+#define MXIC_FAST_DUMMY_CYCLE    0x8
+
+/* ESMT */
+#define ESMT_DUAL_DUMMY_CYCLE    0x4
+#define ESMT_QUAD_DUMMY_CYCLE    0x6
+#define ESMT_FAST_DUMMY_CYCLE    0x8
+
+/* Winbond */
+#define WINBOND_DUAL_DUMMY_CYCLE    0x4
+#define WINBOND_QUAD_DUMMY_CYCLE    0x6
+#define WINBOND_FAST_DUMMY_CYCLE    0x8
+
+/* Spansion */
+#define SPANSION_DUAL_DUMMY_CYCLE    0x4
+#define SPANSION_QUAD_DUMMY_CYCLE    0x6
+#define SPANSION_FAST_DUMMY_CYCLE    0x8
+
+/* GigaDevice */
+#define GIGADEVICE_DUAL_DUMMY_CYCLE    0x4
+#define GIGADEVICE_QUAD_DUMMY_CYCLE    0x6
+#define GIGADEVICE_FAST_DUMMY_CYCLE    0x8
+
+/* EON */
+#define EON_DUAL_DUMMY_CYCLE    0x4
+#define EON_QUAD_DUMMY_CYCLE    0x6
+#define EON_FAST_DUMMY_CYCLE    0x8
+
+/* Used for SST flashes only. */
+#define	OPCODE_BP		0x02	/* Byte program */
+#define	OPCODE_WRDI		0x04	/* Write disable */
+#define	OPCODE_AAI_WP		0xad	/* Auto address increment word program */
+
+/* Used for Macronix flashes only. */
+#define	OPCODE_EN4B		0xb7	/* Enter 4-byte mode */
+#define	OPCODE_EX4B		0xe9	/* Exit 4-byte mode */
+
+/* Used for Spansion flashes only. */
+#define	OPCODE_BRWR		0x17	/* Bank register write */
+
+/* Status Register bits. */
+#define	SR_WIP			1	/* Write in progress */
+#define	SR_WEL			2	/* Write enable latch */
+/* meaning of other SR_* bits may differ between vendors */
+#define	SR_BP0			4	/* Block protect 0 */
+#define	SR_BP1			8	/* Block protect 1 */
+#define	SR_BP2			0x10	/* Block protect 2 */
+#define	SR_SRWD			0x80	/* SR write protect */
+
+#define SR_QUAD_EN_MX           0x40    /* Macronix Quad I/O */
+
+/* Configuration Register bits. */
+#define CR_QUAD_EN_SPAN		0x2     /* Spansion Quad I/O */
+
+/* Define max times to check status register before we give up. */
+#define	MAX_READY_WAIT_JIFFIES	(40 * HZ)	/* M25P16 specs 40s max chip erase */
 #define	MAX_CMD_SIZE		6
+
+#define JEDEC_MFR(_jedec_id)	((_jedec_id) >> 16)
+
+#ifdef CONFIG_M25PXX_PREFER_SMALL_SECTOR_ERASE
+#define PREFER_SMALL_SECTOR_ERASE 1
+#else
+#define PREFER_SMALL_SECTOR_ERASE 0
+#endif
+
+/****************************************************************************/
+/* Extend flash multi-channel read/write type */
+enum m25p80_rd_multi_type {
+	RD_MULTI_NONE = 0x00,
+	RD_DUAL_O     = 0x01,
+	RD_DUAL_IO    = 0x02,
+	RD_QUAD_O     = 0x03,
+	RD_QUAD_IO    = 0x04
+};
+
+enum m25p80_wr_multi_type {
+	WR_MULTI_NONE = 0x00,
+	WR_DUAL_I     = 0x01,
+	WR_DUAL_II    = 0x02,
+	WR_QUAD_I     = 0x03,
+	WR_QUAD_II    = 0x04
+};
+
+struct flash_cmd {
+	volatile uint8_t ppx2_i;	/* flash_cmd; write dual channels */
+	volatile uint8_t ppx2_ii;	/* flash_cmd; write dual channels */
+	volatile uint8_t ppx4_i;	/* flash_cmd; wirte quad channels */
+	volatile uint8_t ppx4_ii;	/* flash_cmd; wirte quad channels */
+	volatile uint8_t readx2_o;	/* flash_cmd; read  dual channels */
+	volatile uint8_t readx2_io;	/* flash_cmd; read  dual channels */
+	volatile uint8_t readx4_o;	/* flash_cmd; read  quad channels */
+	volatile uint8_t readx4_io;	/* flash_cmd; read  quad channels */
+};
+
+struct flash_dummy_cycles_info {
+	uint32_t  rd_dual_dummy;
+	uint32_t  rd_quad_dummy;
+	uint32_t  fast_rd_dummy;
+};
+
+struct flash_rw_multi_type_info {
+	enum m25p80_rd_multi_type rd_dual_type;
+	enum m25p80_rd_multi_type rd_quad_type;
+	enum m25p80_wr_multi_type wr_dual_type;
+	enum m25p80_wr_multi_type wr_quad_type;
+};
+
+struct flash_vendor_info {
+	uint8_t	flash_id;
+	char	vendor_name[16];
+	struct	flash_cmd *cmd;
+	struct	flash_dummy_cycles_info *dummy;
+	struct	flash_rw_multi_type_info *type;
+};
+/* Flash device cmd,
+ * If you want to add new flash vendor
+ * Please add new flash vendor command
+ */
+struct flash_cmd mxic_cmd = {
+	PPX2_I,
+	PPX2_II,
+	PPX4_I,
+	PPX4_II,
+	READX2_I,
+	READX2_IO,
+	READX4_I,
+	READX4_IO,
+};
+
+struct flash_cmd esmt_cmd = {
+	PPX2_I,
+	PPX2_II,
+	PPX4_I,
+	PPX4_II,
+	READX2_I,
+	READX2_IO,
+	READX4_I,
+	READX4_IO,
+};
+
+struct flash_cmd winbond_cmd = {
+	PPX2_I,
+	PPX2_II,
+	PPX4_I,
+	PPX4_II,
+	READX2_I,
+	READX2_IO,
+	READX4_I,
+	READX4_IO,
+};
+
+struct flash_cmd spansion_cmd = {
+	PPX2_I,
+	PPX2_II,
+	PPX4_I,
+	PPX4_II,
+	READX2_I,
+	READX2_IO,
+	READX4_I,
+	READX4_IO,
+};
+
+struct flash_cmd gigadevice_cmd = {
+	PPX2_I,
+	PPX2_II,
+	PPX4_I,
+	PPX4_II,
+	READX2_I,
+	READX2_IO,
+	READX4_I,
+	READX4_IO,
+};
+
+struct flash_cmd eon_cmd = {
+	PPX2_I,
+	PPX2_II,
+	PPX4_I,
+	PPX4_II,
+	READX2_I,
+	READX2_IO,
+	READX4_I,
+	READX4_IO,
+};
+
+/* Flash device dummy info,
+ * If you want to add new flash vendor
+ * Please add new flash vendor dummy
+ */
+struct flash_dummy_cycles_info	mxic_dummy_cycles_info = {
+	MXIC_DUAL_DUMMY_CYCLE,
+	MXIC_QUAD_DUMMY_CYCLE,
+	MXIC_FAST_DUMMY_CYCLE,
+};
+
+struct flash_dummy_cycles_info	esmt_dummy_cycles_info = {
+	ESMT_DUAL_DUMMY_CYCLE,
+	ESMT_QUAD_DUMMY_CYCLE,
+	ESMT_FAST_DUMMY_CYCLE,
+};
+
+struct flash_dummy_cycles_info	winbond_dummy_cycles_info = {
+	WINBOND_DUAL_DUMMY_CYCLE,
+	WINBOND_QUAD_DUMMY_CYCLE,
+	WINBOND_FAST_DUMMY_CYCLE,
+};
+
+struct flash_dummy_cycles_info	spansion_dummy_cycles_info = {
+	SPANSION_DUAL_DUMMY_CYCLE,
+	SPANSION_QUAD_DUMMY_CYCLE,
+	SPANSION_FAST_DUMMY_CYCLE,
+};
+
+struct flash_dummy_cycles_info	gigadevice_dummy_cycles_info = {
+	GIGADEVICE_DUAL_DUMMY_CYCLE,
+	GIGADEVICE_QUAD_DUMMY_CYCLE,
+	GIGADEVICE_FAST_DUMMY_CYCLE,
+};
+
+struct flash_dummy_cycles_info	eon_dummy_cycles_info = {
+	EON_DUAL_DUMMY_CYCLE,
+	EON_QUAD_DUMMY_CYCLE,
+	EON_FAST_DUMMY_CYCLE,
+};
+
+/* Flash device read/write multi type,
+ * If you want to add new flash vendor
+ * Please add new flash vendor dummy
+ */
+struct flash_rw_multi_type_info mxic_rw_multi_type_info = {
+	RD_DUAL_IO,
+	RD_QUAD_IO,
+	WR_MULTI_NONE,
+	WR_QUAD_II
+};
+
+struct flash_rw_multi_type_info esmt_rw_multi_type_info = {
+	RD_DUAL_IO,
+	RD_QUAD_IO,
+	WR_MULTI_NONE,
+	WR_QUAD_II
+};
+
+struct flash_rw_multi_type_info winbond_rw_multi_type_info = {
+	RD_DUAL_IO,
+	RD_QUAD_IO,
+	WR_MULTI_NONE,
+	WR_QUAD_II
+};
+
+struct flash_rw_multi_type_info spansion_rw_multi_type_info = {
+	RD_DUAL_IO,
+	RD_QUAD_IO,
+	WR_MULTI_NONE,
+	WR_QUAD_II
+};
+
+struct flash_rw_multi_type_info gigadevice_rw_multi_type_info = {
+	RD_DUAL_IO,
+	RD_QUAD_IO,
+	WR_MULTI_NONE,
+	WR_QUAD_II
+};
+
+struct flash_rw_multi_type_info eon_rw_multi_type_info = {
+	RD_DUAL_IO,
+	RD_QUAD_IO,
+	WR_MULTI_NONE,
+	WR_QUAD_II
+};
+
+/*-----------------------------------------------------------------*/
+
+struct device_info {
+	uint32_t device_id;
+	struct flash_cmd    *cmd;
+	struct flash_dummy_cycles_info *dummy;
+	struct flash_rw_multi_type_info *type;
+};
+
+/* Add new flash vendor
+ * format (FlashID, vendor name, flash cmd,
+ * flash dummy cycle and read/write multi type),
+ * if you want to add new flash
+ * vendor, please add here.
+ */
+struct flash_vendor_info info_mxic = {
+	0xC2,
+	"MXIC",
+	&mxic_cmd,
+	&mxic_dummy_cycles_info,
+	&mxic_rw_multi_type_info
+};
+
+struct flash_vendor_info info_esmt = {
+	0x8C,
+	"ESMT",
+	&esmt_cmd,
+	&esmt_dummy_cycles_info,
+	&esmt_rw_multi_type_info
+};
+
+struct flash_vendor_info info_winbond = {
+	0xEF,
+	"Winbond",
+	&winbond_cmd,
+	&winbond_dummy_cycles_info,
+	&winbond_rw_multi_type_info
+};
+
+struct flash_vendor_info info_spansion = {
+	0x01,
+	"Spansion",
+	&spansion_cmd,
+	&spansion_dummy_cycles_info,
+	&spansion_rw_multi_type_info
+};
+
+struct flash_vendor_info info_gigadevice = {
+	0xC8,
+	"GigaDevice",
+	&gigadevice_cmd,
+	&gigadevice_dummy_cycles_info,
+	&gigadevice_rw_multi_type_info
+};
+
+struct flash_vendor_info info_eon = {
+	0x1C,
+	"EON",
+	&eon_cmd,
+	&eon_dummy_cycles_info,
+	&eon_rw_multi_type_info
+};
+
+/* Flash vendors eg: MXIC.... */
+struct flash_device {
+	struct flash_vendor_info *vendors[6];
+};
+
+struct flash_device device = {
+	{ &info_mxic, &info_esmt, &info_winbond, &info_spansion, &info_gigadevice, &info_eon }
+};
+
+enum read_type {
+	M25P80_NORMAL = 0,
+	M25P80_FAST,
+	M25P80_QUAD,
+	M25P80_DUAL,
+	M25P80_AUTO
+};
+
+enum write_type {
+	M25P80_QUAD_WRITE = 5,
+	M25P80_DUAL_WRITE,
+	M25P80_NORMAL_WRITE,
+	M25P80_AUTO_WRITE
+};
+
 struct m25p {
	struct spi_device	*spi;
-	struct spi_nor		spi_nor;
+	struct mutex		lock;
	struct mtd_info		mtd;
-	u8			command[MAX_CMD_SIZE];
+	u16			page_size;
+	u16			addr_width;
+	u8			erase_opcode;
+	u8			read_opcode;
+	u8			program_opcode;
+	u8			*command;
+	enum read_type		flash_read;
+	/* support special flash write mode */
+	enum write_type		flash_write;
+	bool			fast_read;
+	/* support multi-channel write and read */
+	bool			quad;
+	bool			dual;
+	/* support auto mode */
+	bool			auto_mode;
+	/* support flash vendor information */
+	struct	device_info	dev_info;
+	/* flash read dummy cycles */
+	unsigned int	dummy;
+	/* support flash read/write multi-channel type */
+	unsigned int	write_type;
+	unsigned int	read_type;
 };

-static int m25p80_read_reg(struct spi_nor *nor, u8 code, u8 *val, int len)
+static inline struct m25p *mtd_to_m25p(struct mtd_info *mtd)
 {
-	struct m25p *flash = nor->priv;
-	struct spi_device *spi = flash->spi;
+	return container_of(mtd, struct m25p, mtd);
+}
+
+/****************************************************************************/
+
+/*
+ * Internal helper functions
+ */
+
+/*
+ * Read the status register, returning its value in the location
+ * Return the status register value.
+ * Returns negative if error occurred.
+ */
+static int read_sr(struct m25p *flash)
+{
+	ssize_t retval;
+	u8 code = OPCODE_RDSR;
+	u8 val;
+
+	retval = spi_write_then_read(flash->spi, &code, 1, &val, 1);
+
+	if (retval < 0) {
+		dev_err(&flash->spi->dev, "error %d reading SR\n",
+				(int) retval);
+		return retval;
+	}
+
+	return val;
+}
+
+/*
+ * Read configuration register, returning its value in the
+ * location. Return the configuration register value.
+ * Returns negative if error occured.
+ */
+static int read_cr(struct m25p *flash)
+{
+	u8 code = OPCODE_RDCR;
	int ret;
+	u8 val;

-	ret = spi_write_then_read(spi, &code, 1, val, len);
-	if (ret < 0)
-		dev_err(&spi->dev, "error %d reading %x\n", ret, code);
+	ret = spi_write_then_read(flash->spi, &code, 1, &val, 1);
+	if (ret < 0) {
+		dev_err(&flash->spi->dev, "error %d reading CR\n", ret);
+		return ret;
+	}

-	return ret;
+	return val;
 }

-static void m25p_addr2cmd(struct spi_nor *nor, unsigned int addr, u8 *cmd)
+/*
+ * Write status register 1 byte
+ * Returns negative if error occurred.
+ */
+static int write_sr(struct m25p *flash, u8 val)
 {
-	/* opcode is in cmd[0] */
-	cmd[1] = addr >> (nor->addr_width * 8 -  8);
-	cmd[2] = addr >> (nor->addr_width * 8 - 16);
-	cmd[3] = addr >> (nor->addr_width * 8 - 24);
-	cmd[4] = addr >> (nor->addr_width * 8 - 32);
+	flash->command[0] = OPCODE_WRSR;
+	flash->command[1] = val;
+
+	return spi_write(flash->spi, flash->command, 2);
 }

-static int m25p_cmdsz(struct spi_nor *nor)
+/*
+ * Set write enable latch with Write Enable command.
+ * Returns negative if error occurred.
+ */
+static inline int write_enable(struct m25p *flash)
 {
-	return 1 + nor->addr_width;
+	u8	code = OPCODE_WREN;
+
+	return spi_write_then_read(flash->spi, &code, 1, NULL, 0);
 }

-static int m25p80_write_reg(struct spi_nor *nor, u8 opcode, u8 *buf, int len,
-			int wr_en)
+/*
+ * Send write disble instruction to the chip.
+ */
+static inline int write_disable(struct m25p *flash)
 {
-	struct m25p *flash = nor->priv;
-	struct spi_device *spi = flash->spi;
+	u8	code = OPCODE_WRDI;

-	flash->command[0] = opcode;
-	if (buf)
-		memcpy(&flash->command[1], buf, len);
+	return spi_write_then_read(flash->spi, &code, 1, NULL, 0);
+}
+
+/*
+ * Enable/disable 4-byte addressing mode.
+ */
+static inline int set_4byte(struct m25p *flash, u32 jedec_id, int enable)
+{
+	int status;
+	bool need_wren = false;
+
+	switch (JEDEC_MFR(jedec_id)) {
+	case CFI_MFR_ST: /* Micron, actually */
+		/* Some Micron need WREN command; all will accept it */
+		need_wren = true;
+	case CFI_MFR_MACRONIX:
+	case 0xEF /* winbond */:
+		if (need_wren)
+			write_enable(flash);

-	return spi_write(spi, flash->command, len + 1);
+		flash->command[0] = enable ? OPCODE_EN4B : OPCODE_EX4B;
+		status = spi_write(flash->spi, flash->command, 1);
+
+		if (need_wren)
+			write_disable(flash);
+
+		return status;
+	default:
+		/* Spansion style */
+		flash->command[0] = OPCODE_BRWR;
+		flash->command[1] = enable << 7;
+		return spi_write(flash->spi, flash->command, 2);
+	}
 }

-static void m25p80_write(struct spi_nor *nor, loff_t to, size_t len,
-			size_t *retlen, const u_char *buf)
+/*
+ * Service routine to read status register until ready, or timeout occurs.
+ * Returns non-zero if error.
+ */
+static int wait_till_ready(struct m25p *flash)
 {
-	struct m25p *flash = nor->priv;
-	struct spi_device *spi = flash->spi;
-	struct spi_transfer t[2] = {};
-	struct spi_message m;
-	int cmd_sz = m25p_cmdsz(nor);
+	unsigned long deadline;
+	int sr;

-	spi_message_init(&m);
+	deadline = jiffies + MAX_READY_WAIT_JIFFIES;

-	if (nor->program_opcode == SPINOR_OP_AAI_WP && nor->sst_write_second)
-		cmd_sz = 1;
+	do {
+		if ((sr = read_sr(flash)) < 0)
+			break;
+		else if (!(sr & SR_WIP))
+			return 0;

-	flash->command[0] = nor->program_opcode;
-	m25p_addr2cmd(nor, to, flash->command);
+		cond_resched();

-	t[0].tx_buf = flash->command;
-	t[0].len = cmd_sz;
-	spi_message_add_tail(&t[0], &m);
+	} while (!time_after_eq(jiffies, deadline));

-	t[1].tx_buf = buf;
-	t[1].len = len;
-	spi_message_add_tail(&t[1], &m);
+	return 1;
+}
+
+/*
+ * Write status Register and configuration register with 2 bytes
+ * The first byte will be written to the status register, while the
+ * second byte will be written to the configuration register.
+ * Return negative if error occured.
+ */
+static int write_sr_cr(struct m25p *flash, u16 val)
+{
+	flash->command[0] = OPCODE_WRSR;
+	flash->command[1] = val & 0xff;
+	flash->command[2] = (val >> 8);
+
+	return spi_write(flash->spi, flash->command, 3);
+}
+
+static int macronix_quad_enable(struct m25p *flash)
+{
+	int ret, val;
+	u8 cmd[2];
+	cmd[0] = OPCODE_WRSR;
+
+	val = read_sr(flash);
+	cmd[1] = val | SR_QUAD_EN_MX;
+	write_enable(flash);
+
+	spi_write(flash->spi, &cmd, 2);
+
+	if (wait_till_ready(flash))
+		return 1;
+
+	ret = read_sr(flash);
+	if (!(ret > 0 && (ret & SR_QUAD_EN_MX))) {
+		dev_err(&flash->spi->dev, "Macronix Quad bit not set\n");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int spansion_quad_enable(struct m25p *flash)
+{
+	int ret;
+	int quad_en = CR_QUAD_EN_SPAN << 8;
+
+	write_enable(flash);
+
+	ret = write_sr_cr(flash, quad_en);
+	if (ret < 0) {
+		dev_err(&flash->spi->dev,
+			"error while writing configuration register\n");
+		return -EINVAL;
+	}
+
+	/* read back and check it */
+	ret = read_cr(flash);
+	if (!(ret > 0 && (ret & CR_QUAD_EN_SPAN))) {
+		dev_err(&flash->spi->dev, "Spansion Quad bit not set\n");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int set_quad_mode(struct m25p *flash, u32 jedec_id)
+{
+	int status;
+
+	switch (JEDEC_MFR(jedec_id)) {
+	case CFI_MFR_MACRONIX:
+		status = macronix_quad_enable(flash);
+		if (status) {
+			dev_err(&flash->spi->dev,
+				"Macronix quad-read not enabled\n");
+			return -EINVAL;
+		}
+		return status;
+	default:
+		status = spansion_quad_enable(flash);
+		if (status) {
+			dev_err(&flash->spi->dev,
+				"Spansion quad-read not enabled\n");
+			return -EINVAL;
+		}
+		return status;
+	}
+}
+
+/* Get flash device information such as command, dummy cycle and type */
+static int get_flash_device(struct m25p *flash, u32 jedec_id)
+{
+	uint32_t i;
+	uint32_t size;
+
+	flash->dev_info.device_id = JEDEC_MFR(jedec_id);
+	size = sizeof(struct flash_device)/sizeof(struct flash_vendor_info *);
+	for (i = 0; i < size; i++) {
+		if (flash->dev_info.device_id == device.vendors[i]->flash_id) {
+			flash->dev_info.cmd = device.vendors[i]->cmd;
+			flash->dev_info.dummy = device.vendors[i]->dummy;
+			flash->dev_info.type = device.vendors[i]->type;
+			printk(KERN_INFO "flash vendor: %s\n", device.vendors[i]->vendor_name);
+			return 0;
+		}
+	}
+	printk(KERN_INFO "\n>>>>>No Flash Vendor support (0x%x)<<<<<\n\n", jedec_id);
+#if 1
+	printk("use MXIC as flash vendor instead\n");
+	flash->dev_info.cmd = device.vendors[0]->cmd;
+	flash->dev_info.dummy = device.vendors[0]->dummy;
+	flash->dev_info.type = device.vendors[0]->type;
+	return 0;
+#else
+	return -EINVAL;
+#endif
+}
+/*
+ * Erase the whole flash memory
+ *
+ * Returns 0 if successful, non-zero otherwise.
+ */
+static int erase_chip(struct m25p *flash)
+{
+	pr_debug("%s: %s %lldKiB\n", dev_name(&flash->spi->dev), __func__,
+			(long long)(flash->mtd.size >> 10));
+
+	/* Wait until finished previous write command. */
+	if (wait_till_ready(flash))
+		return 1;
+
+	/* Send write enable, then erase commands. */
+	write_enable(flash);
+
+	/* Set up command buffer. */
+	flash->command[0] = OPCODE_CHIP_ERASE;
+
+	spi_write(flash->spi, flash->command, 1);
+
+	return 0;
+}

-	spi_sync(spi, &m);
+static void m25p_addr2cmd(struct m25p *flash, unsigned int addr, u8 *cmd)
+{
+	/* opcode is in cmd[0] */
+	cmd[1] = addr >> (flash->addr_width * 8 -  8);
+	cmd[2] = addr >> (flash->addr_width * 8 - 16);
+	cmd[3] = addr >> (flash->addr_width * 8 - 24);
+	cmd[4] = addr >> (flash->addr_width * 8 - 32);
+}

-	*retlen += m.actual_length - cmd_sz;
+static int m25p_cmdsz(struct m25p *flash)
+{
+	return 1 + flash->addr_width;
 }

-static inline unsigned int m25p80_rx_nbits(struct spi_nor *nor)
+/*
+ * Erase one sector of flash memory at offset ``offset'' which is any
+ * address within the sector which should be erased.
+ *
+ * Returns 0 if successful, non-zero otherwise.
+ */
+static int erase_sector(struct m25p *flash, u32 offset)
 {
-	switch (nor->flash_read) {
-	case SPI_NOR_DUAL:
-		return 2;
-	case SPI_NOR_QUAD:
+	pr_debug("%s: %s %dKiB at 0x%08x\n", dev_name(&flash->spi->dev),
+			__func__, flash->mtd.erasesize / 1024, offset);
+
+	/* Wait until finished previous write command. */
+	if (wait_till_ready(flash))
+		return 1;
+
+	/* Send write enable, then erase commands. */
+	write_enable(flash);
+
+	/* Set up command buffer. */
+	flash->command[0] = flash->erase_opcode;
+	m25p_addr2cmd(flash, offset, flash->command);
+
+	spi_write(flash->spi, flash->command, m25p_cmdsz(flash));
+
+	return 0;
+}
+
+/****************************************************************************/
+
+/*
+ * MTD implementation
+ */
+
+/*
+ * Erase an address range on the flash chip.  The address range may extend
+ * one or more erase sectors.  Return an error is there is a problem erasing.
+ */
+static int m25p80_erase(struct mtd_info *mtd, struct erase_info *instr)
+{
+	struct m25p *flash = mtd_to_m25p(mtd);
+	u32 addr,len;
+	uint32_t rem;
+
+	pr_debug("%s: %s at 0x%llx, len %lld\n", dev_name(&flash->spi->dev),
+			__func__, (long long)instr->addr,
+			(long long)instr->len);
+
+	div_u64_rem(instr->len, mtd->erasesize, &rem);
+	if (rem)
+		return -EINVAL;
+
+	addr = instr->addr;
+	len = instr->len;
+
+	mutex_lock(&flash->lock);
+
+	/* whole-chip erase? */
+	if (len == flash->mtd.size) {
+		if (erase_chip(flash)) {
+			instr->state = MTD_ERASE_FAILED;
+			mutex_unlock(&flash->lock);
+			return -EIO;
+		}
+
+	/* REVISIT in some cases we could speed up erasing large regions
+	 * by using OPCODE_SE instead of OPCODE_BE_4K.  We may have set up
+	 * to use "small sector erase", but that's not always optimal.
+	 */
+
+	/* "sector"-at-a-time erase */
+	} else {
+		while (len) {
+			if (erase_sector(flash, addr)) {
+				instr->state = MTD_ERASE_FAILED;
+				mutex_unlock(&flash->lock);
+				return -EIO;
+			}
+
+			addr += mtd->erasesize;
+			len -= mtd->erasesize;
+		}
+	}
+
+	mutex_unlock(&flash->lock);
+
+	instr->state = MTD_ERASE_DONE;
+	mtd_erase_callback(instr);
+
+	return 0;
+}
+
+/*
+ * Dummy Cycle calculation for different type of read.
+ * It can be used to support more commands with
+ * different dummy cycle requirements.
+ */
+static inline int m25p80_dummy_cycles_read(struct m25p *flash)
+{
+	switch (flash->flash_read) {
+	case M25P80_FAST:
+	case M25P80_DUAL:
+	case M25P80_QUAD:
+		return 1;
+	case M25P80_NORMAL:
+	case M25P80_AUTO:
+		return 0;
+	default:
+		dev_err(&flash->spi->dev, "No valid read type supported\n");
+		return -1;
+	}
+}
+
+static inline unsigned int m25p80_rx_nbits(const struct m25p *flash)
+{
+	switch (flash->flash_read) {
+	case M25P80_QUAD:
		return 4;
	default:
		return 0;
@@ -118,72 +878,636 @@
 }

 /*
- * Read an address range from the nor chip.  The address range
+ * Read an address range from the flash chip.  The address range
  * may be any size provided it is within the physical boundaries.
  */
-static int m25p80_read(struct spi_nor *nor, loff_t from, size_t len,
-			size_t *retlen, u_char *buf)
+static int m25p80_read(struct mtd_info *mtd, loff_t from, size_t len,
+	size_t *retlen, u_char *buf)
 {
-	struct m25p *flash = nor->priv;
-	struct spi_device *spi = flash->spi;
+	struct m25p *flash = mtd_to_m25p(mtd);
	struct spi_transfer t[2];
	struct spi_message m;
-	int dummy = nor->read_dummy;
-	int ret;
+	uint8_t opcode;
+	int dummy;
+	unsigned int type;
+	uint8_t	mode;

-	/* Wait till previous write/erase is done. */
-	ret = nor->wait_till_ready(nor);
-	if (ret)
-		return ret;
+	pr_debug("%s: %s from 0x%08x, len %zd\n", dev_name(&flash->spi->dev),
+			__func__, (u32)from, len);

	spi_message_init(&m);
	memset(t, 0, (sizeof t));

-	flash->command[0] = nor->read_opcode;
-	m25p_addr2cmd(nor, from, flash->command);
+	/* NOTE:
+	 * OPCODE_FAST_READ (if available) is faster.
+	 * Should add 1 byte DUMMY_BYTE.
+	 */

+	dummy =  m25p80_dummy_cycles_read(flash);
+	if (dummy < 0) {
+		dev_err(&flash->spi->dev, "No valid read command supported\n");
+		return -EINVAL;
+	}
	t[0].tx_buf = flash->command;
-	t[0].len = m25p_cmdsz(nor) + dummy;
+	t[0].len = m25p_cmdsz(flash) + dummy;
	spi_message_add_tail(&t[0], &m);

	t[1].rx_buf = buf;
-	t[1].rx_nbits = m25p80_rx_nbits(nor);
+	t[1].rx_nbits = m25p80_rx_nbits(flash);
	t[1].len = len;
	spi_message_add_tail(&t[1], &m);

-	spi_sync(spi, &m);
+	mutex_lock(&flash->lock);
+
+	/* Wait till previous write/erase is done. */
+	if (wait_till_ready(flash)) {
+		/* REVISIT status return?? */
+		mutex_unlock(&flash->lock);
+		return 1;
+	}
+
+	/* FIXME switch to OPCODE_FAST_READ.  It's required for higher
+	 * clocks; and at this writing, every chip this driver handles
+	 * supports that opcode.
+	 */
+
+	/* Set up the write data buffer. */
+	opcode = flash->read_opcode;
+	mode = flash->flash_read;
+	type = flash->read_type;
+	flash->command[0] = opcode;
+	m25p_addr2cmd(flash, from, flash->command);
+	flash->command[5] = mode;
+	flash->command[6] = flash->dummy;
+	flash->command[7] = type;
+
+	spi_sync(flash->spi, &m);
+
+	*retlen = m.actual_length - m25p_cmdsz(flash) - dummy;
+
+	mutex_unlock(&flash->lock);

-	*retlen = m.actual_length - m25p_cmdsz(nor) - dummy;
	return 0;
 }

-static int m25p80_erase(struct spi_nor *nor, loff_t offset)
+/*
+ * Write an address range to the flash chip.  Data must be written in
+ * FLASH_PAGESIZE chunks.  The address range may be any size provided
+ * it is within the physical boundaries.
+ */
+static int m25p80_write(struct mtd_info *mtd, loff_t to, size_t len,
+	size_t *retlen, const u_char *buf)
 {
-	struct m25p *flash = nor->priv;
-	int ret;
+	struct m25p *flash = mtd_to_m25p(mtd);
+	u32 page_offset, page_size;
+	struct spi_transfer t[2];
+	struct spi_message m;
+	unsigned int type;
+	uint8_t mode;
+
+	pr_debug("%s: %s to 0x%08x, len %zd\n", dev_name(&flash->spi->dev),
+			__func__, (u32)to, len);
+
+	spi_message_init(&m);
+	memset(t, 0, (sizeof t));
+
+	t[0].tx_buf = flash->command;
+	t[0].len = m25p_cmdsz(flash);
+	spi_message_add_tail(&t[0], &m);
+
+	t[1].tx_buf = buf;
+	spi_message_add_tail(&t[1], &m);
+
+	mutex_lock(&flash->lock);
+
+	/* Wait until finished previous write command. */
+	if (wait_till_ready(flash)) {
+		mutex_unlock(&flash->lock);
+		return 1;
+	}
+
+	write_enable(flash);
+
+	/* Set up the opcode in the write buffer. */
+	mode = flash->flash_write;
+	type = flash->write_type;
+	flash->command[0] = flash->program_opcode;
+	m25p_addr2cmd(flash, to, flash->command);
+	flash->command[5] = mode;
+	flash->command[7] = type;
+	page_offset = to & (flash->page_size - 1);
+
+	/* do all the bytes fit onto one page? */
+	if (page_offset + len <= flash->page_size) {
+		t[1].len = len;
+
+		spi_sync(flash->spi, &m);
+
+		*retlen = m.actual_length - m25p_cmdsz(flash);
+	} else {
+		u32 i;
+
+		/* the size of data remaining on the first page */
+		page_size = flash->page_size - page_offset;
+
+		t[1].len = page_size;
+		spi_sync(flash->spi, &m);
+
+		*retlen = m.actual_length - m25p_cmdsz(flash);
+
+		/* write everything in flash->page_size chunks */
+		for (i = page_size; i < len; i += page_size) {
+			page_size = len - i;
+			if (page_size > flash->page_size)
+				page_size = flash->page_size;
+
+			/* write the next page to flash */
+			m25p_addr2cmd(flash, to + i, flash->command);
+
+			t[1].tx_buf = buf + i;
+			t[1].len = page_size;
+
+			wait_till_ready(flash);
+
+			write_enable(flash);
+
+			spi_sync(flash->spi, &m);
+
+			*retlen += m.actual_length - m25p_cmdsz(flash);
+		}
+	}
+
+	mutex_unlock(&flash->lock);
+
+	return 0;
+}

-	dev_dbg(nor->dev, "%dKiB at 0x%08x\n",
-		flash->mtd.erasesize / 1024, (u32)offset);
+static int sst_write(struct mtd_info *mtd, loff_t to, size_t len,
+		size_t *retlen, const u_char *buf)
+{
+	struct m25p *flash = mtd_to_m25p(mtd);
+	struct spi_transfer t[2];
+	struct spi_message m;
+	size_t actual;
+	int cmd_sz, ret;
+
+	pr_debug("%s: %s to 0x%08x, len %zd\n", dev_name(&flash->spi->dev),
+			__func__, (u32)to, len);
+
+	spi_message_init(&m);
+	memset(t, 0, (sizeof t));
+
+	t[0].tx_buf = flash->command;
+	t[0].len = m25p_cmdsz(flash);
+	spi_message_add_tail(&t[0], &m);
+
+	t[1].tx_buf = buf;
+	spi_message_add_tail(&t[1], &m);
+
+	mutex_lock(&flash->lock);

	/* Wait until finished previous write command. */
-	ret = nor->wait_till_ready(nor);
+	ret = wait_till_ready(flash);
	if (ret)
-		return ret;
+		goto time_out;

-	/* Send write enable, then erase commands. */
-	ret = nor->write_reg(nor, SPINOR_OP_WREN, NULL, 0, 0);
+	write_enable(flash);
+
+	actual = to % 2;
+	/* Start write from odd address. */
+	if (actual) {
+		flash->command[0] = OPCODE_BP;
+		m25p_addr2cmd(flash, to, flash->command);
+
+		/* write one byte. */
+		t[1].len = 1;
+		spi_sync(flash->spi, &m);
+		ret = wait_till_ready(flash);
+		if (ret)
+			goto time_out;
+		*retlen += m.actual_length - m25p_cmdsz(flash);
+	}
+	to += actual;
+
+	flash->command[0] = OPCODE_AAI_WP;
+	m25p_addr2cmd(flash, to, flash->command);
+
+	/* Write out most of the data here. */
+	cmd_sz = m25p_cmdsz(flash);
+	for (; actual < len - 1; actual += 2) {
+		t[0].len = cmd_sz;
+		/* write two bytes. */
+		t[1].len = 2;
+		t[1].tx_buf = buf + actual;
+
+		spi_sync(flash->spi, &m);
+		ret = wait_till_ready(flash);
+		if (ret)
+			goto time_out;
+		*retlen += m.actual_length - cmd_sz;
+		cmd_sz = 1;
+		to += 2;
+	}
+	write_disable(flash);
+	ret = wait_till_ready(flash);
	if (ret)
-		return ret;
+		goto time_out;

-	/* Set up command buffer. */
-	flash->command[0] = nor->erase_opcode;
-	m25p_addr2cmd(nor, offset, flash->command);
+	/* Write out trailing byte if it exists. */
+	if (actual != len) {
+		write_enable(flash);
+		flash->command[0] = OPCODE_BP;
+		m25p_addr2cmd(flash, to, flash->command);
+		t[0].len = m25p_cmdsz(flash);
+		t[1].len = 1;
+		t[1].tx_buf = buf + actual;
+
+		spi_sync(flash->spi, &m);
+		ret = wait_till_ready(flash);
+		if (ret)
+			goto time_out;
+		*retlen += m.actual_length - m25p_cmdsz(flash);
+		write_disable(flash);
+	}

-	spi_write(flash->spi, flash->command, m25p_cmdsz(nor));
+time_out:
+	mutex_unlock(&flash->lock);
+	return ret;
+}

-	return 0;
+static int m25p80_lock(struct mtd_info *mtd, loff_t ofs, uint64_t len)
+{
+	struct m25p *flash = mtd_to_m25p(mtd);
+	uint32_t offset = ofs;
+	uint8_t status_old, status_new;
+	int res = 0;
+
+	mutex_lock(&flash->lock);
+	/* Wait until finished previous command */
+	if (wait_till_ready(flash)) {
+		res = 1;
+		goto err;
+	}
+
+	status_old = read_sr(flash);
+
+	if (offset < flash->mtd.size-(flash->mtd.size/2))
+		status_new = status_old | SR_BP2 | SR_BP1 | SR_BP0;
+	else if (offset < flash->mtd.size-(flash->mtd.size/4))
+		status_new = (status_old & ~SR_BP0) | SR_BP2 | SR_BP1;
+	else if (offset < flash->mtd.size-(flash->mtd.size/8))
+		status_new = (status_old & ~SR_BP1) | SR_BP2 | SR_BP0;
+	else if (offset < flash->mtd.size-(flash->mtd.size/16))
+		status_new = (status_old & ~(SR_BP0|SR_BP1)) | SR_BP2;
+	else if (offset < flash->mtd.size-(flash->mtd.size/32))
+		status_new = (status_old & ~SR_BP2) | SR_BP1 | SR_BP0;
+	else if (offset < flash->mtd.size-(flash->mtd.size/64))
+		status_new = (status_old & ~(SR_BP2|SR_BP0)) | SR_BP1;
+	else
+		status_new = (status_old & ~(SR_BP2|SR_BP1)) | SR_BP0;
+
+	/* Only modify protection if it will not unlock other areas */
+	if ((status_new&(SR_BP2|SR_BP1|SR_BP0)) >
+					(status_old&(SR_BP2|SR_BP1|SR_BP0))) {
+		write_enable(flash);
+		if (write_sr(flash, status_new) < 0) {
+			res = 1;
+			goto err;
+		}
+	}
+
+err:	mutex_unlock(&flash->lock);
+	return res;
+}
+
+static int m25p80_unlock(struct mtd_info *mtd, loff_t ofs, uint64_t len)
+{
+	struct m25p *flash = mtd_to_m25p(mtd);
+	uint32_t offset = ofs;
+	uint8_t status_old, status_new;
+	int res = 0;
+
+	mutex_lock(&flash->lock);
+	/* Wait until finished previous command */
+	if (wait_till_ready(flash)) {
+		res = 1;
+		goto err;
+	}
+
+	status_old = read_sr(flash);
+
+	if (offset+len > flash->mtd.size-(flash->mtd.size/64))
+		status_new = status_old & ~(SR_BP2|SR_BP1|SR_BP0);
+	else if (offset+len > flash->mtd.size-(flash->mtd.size/32))
+		status_new = (status_old & ~(SR_BP2|SR_BP1)) | SR_BP0;
+	else if (offset+len > flash->mtd.size-(flash->mtd.size/16))
+		status_new = (status_old & ~(SR_BP2|SR_BP0)) | SR_BP1;
+	else if (offset+len > flash->mtd.size-(flash->mtd.size/8))
+		status_new = (status_old & ~SR_BP2) | SR_BP1 | SR_BP0;
+	else if (offset+len > flash->mtd.size-(flash->mtd.size/4))
+		status_new = (status_old & ~(SR_BP0|SR_BP1)) | SR_BP2;
+	else if (offset+len > flash->mtd.size-(flash->mtd.size/2))
+		status_new = (status_old & ~SR_BP1) | SR_BP2 | SR_BP0;
+	else
+		status_new = (status_old & ~SR_BP0) | SR_BP2 | SR_BP1;
+
+	/* Only modify protection if it will not lock other areas */
+	if ((status_new&(SR_BP2|SR_BP1|SR_BP0)) <
+					(status_old&(SR_BP2|SR_BP1|SR_BP0))) {
+		write_enable(flash);
+		if (write_sr(flash, status_new) < 0) {
+			res = 1;
+			goto err;
+		}
+	}
+
+err:	mutex_unlock(&flash->lock);
+	return res;
+}
+
+/****************************************************************************/
+
+/*
+ * SPI device driver setup and teardown
+ */
+
+struct flash_info {
+	/* JEDEC id zero means "no ID" (most older chips); otherwise it has
+	 * a high byte of zero plus three data bytes: the manufacturer id,
+	 * then a two byte device id.
+	 */
+	u32		jedec_id;
+	u16             ext_id;
+
+	/* The size listed here is what works with OPCODE_SE, which isn't
+	 * necessarily called a "sector" by the vendor.
+	 */
+	unsigned	sector_size;
+	u16		n_sectors;
+
+	u16		page_size;
+	u16		addr_width;
+
+	u16		flags;
+#define	SECT_4K		0x01		/* OPCODE_BE_4K works uniformly */
+#define	M25P_NO_ERASE	0x02		/* No erase command needed */
+#define	SST_WRITE	0x04		/* use SST byte programming */
+#define	M25P_NO_FR	0x08		/* Can't do fastread */
+#define	SECT_4K_PMC	0x10		/* OPCODE_BE_4K_PMC works uniformly */
+#define	M25P80_QUAD_READ	0x20    /* Flash supports Quad Read */
+};
+
+#define INFO(_jedec_id, _ext_id, _sector_size, _n_sectors, _flags)	\
+	((kernel_ulong_t)&(struct flash_info) {				\
+		.jedec_id = (_jedec_id),				\
+		.ext_id = (_ext_id),					\
+		.sector_size = (_sector_size),				\
+		.n_sectors = (_n_sectors),				\
+		.page_size = 256,					\
+		.flags = (_flags),					\
+	})
+
+#define CAT25_INFO(_sector_size, _n_sectors, _page_size, _addr_width, _flags)	\
+	((kernel_ulong_t)&(struct flash_info) {				\
+		.sector_size = (_sector_size),				\
+		.n_sectors = (_n_sectors),				\
+		.page_size = (_page_size),				\
+		.addr_width = (_addr_width),				\
+		.flags = (_flags),					\
+	})
+
+/* NOTE: double check command sets and memory organization when you add
+ * more flash chips.  This current list focusses on newer chips, which
+ * have been converging on command sets which including JEDEC ID.
+ */
+static const struct spi_device_id m25p_ids[] = {
+	/* Atmel -- some are (confusingly) marketed as "DataFlash" */
+	{ "at25fs010",  INFO(0x1f6601, 0, 32 * 1024,   4, SECT_4K) },
+	{ "at25fs040",  INFO(0x1f6604, 0, 64 * 1024,   8, SECT_4K) },
+
+	{ "at25df041a", INFO(0x1f4401, 0, 64 * 1024,   8, SECT_4K) },
+	{ "at25df321a", INFO(0x1f4701, 0, 64 * 1024,  64, SECT_4K) },
+	{ "at25df641",  INFO(0x1f4800, 0, 64 * 1024, 128, SECT_4K) },
+
+	{ "at26f004",   INFO(0x1f0400, 0, 64 * 1024,  8, SECT_4K) },
+	{ "at26df081a", INFO(0x1f4501, 0, 64 * 1024, 16, SECT_4K) },
+	{ "at26df161a", INFO(0x1f4601, 0, 64 * 1024, 32, SECT_4K) },
+	{ "at26df321",  INFO(0x1f4700, 0, 64 * 1024, 64, SECT_4K) },
+
+	{ "at45db081d", INFO(0x1f2500, 0, 64 * 1024, 16, SECT_4K) },
+
+	/* EON -- en25xxx */
+	{ "en25f16",    INFO(0x1c3115, 0, 64 * 1024,   32, SECT_4K) },
+	{ "en25f32", INFO(0x1c3116, 0, 64 * 1024,  64, SECT_4K) },
+	{ "en25p32", INFO(0x1c2016, 0, 64 * 1024,  64, 0) },
+	{ "en25q16",   INFO(0x1c3015, 0, 64 * 1024,   32, 0) },
+	{ "en25q32b", INFO(0x1c3016, 0, 64 * 1024,  64, 0) },
+	{ "en25p64", INFO(0x1c2017, 0, 64 * 1024, 128, 0) },
+	{ "en25q64", INFO(0x1c3017, 0, 64 * 1024, 128, SECT_4K) },
+	{ "en25qh128", INFO(0x1c7018, 0, 64 * 1024, 256, 0) },
+	{ "en25qh256", INFO(0x1c7019, 0, 64 * 1024, 512, 0) },
+
+	/* ESMT */
+	{ "f25l08qa(2s)", INFO(0x8c4014, 0, 64 * 1024, 16, SECT_4K) },
+	{ "f25l16pa(2s)", INFO(0x8c2115, 0, 64 * 1024, 32, SECT_4K) },
+	{ "f25l16qa(2s)", INFO(0x8c4015, 0, 64 * 1024, 32, SECT_4K) },
+	{ "f25l32pa", INFO(0x8c2016, 0, 64 * 1024, 64, SECT_4K) },
+	{ "f25l32qa", INFO(0x8c4016, 0, 64 * 1024, 64, SECT_4K) },
+	{ "f25l32qa(2s)", INFO(0x8c4116, 0, 64 * 1024, 64, SECT_4K) },
+	{ "f25l64qa", INFO(0x8c4117, 0, 64 * 1024, 128, SECT_4K) },
+	{ "f25l128qa", INFO(0x8c4118, 0, 64 * 1024, 256, SECT_4K) },
+
+	/* Everspin */
+	{ "mr25h256", CAT25_INFO( 32 * 1024, 1, 256, 2, M25P_NO_ERASE | M25P_NO_FR) },
+	{ "mr25h10",  CAT25_INFO(128 * 1024, 1, 256, 3, M25P_NO_ERASE | M25P_NO_FR) },
+
+	/* GigaDevice */
+	{ "gd25q80", INFO(0xc84014, 0, 64 * 1024,  16, SECT_4K) },
+	{ "gd25q16", INFO(0xc84015, 0, 64 * 1024,  32, SECT_4K) },
+	{ "gd25q32", INFO(0xc84016, 0, 64 * 1024,  64, SECT_4K) },
+	{ "gd25q64", INFO(0xc84017, 0, 64 * 1024, 128, SECT_4K) },
+	{ "gd25q128", INFO(0xc84018, 0, 64 * 1024, 256, SECT_4K) },
+
+	/* Intel/Numonyx -- xxxs33b */
+	{ "160s33b",  INFO(0x898911, 0, 64 * 1024,  32, 0) },
+	{ "320s33b",  INFO(0x898912, 0, 64 * 1024,  64, 0) },
+	{ "640s33b",  INFO(0x898913, 0, 64 * 1024, 128, 0) },
+
+	/* Macronix */
+	{ "mx25l2005a",  INFO(0xc22012, 0, 64 * 1024,   4, SECT_4K) },
+	{ "mx25l4005a",  INFO(0xc22013, 0, 64 * 1024,   8, SECT_4K) },
+	{ "mx25l8005",   INFO(0xc22014, 0, 64 * 1024,  16, 0) },
+	{ "mx25l1606e",  INFO(0xc22015, 0, 64 * 1024,  32, SECT_4K) },
+	{ "mx25l1633e",  INFO(0xc22415, 0, 64 * 1024,  32, SECT_4K) },
+	{ "mx25l1635e",  INFO(0xc22515, 0, 64 * 1024,  32, SECT_4K) },
+	{ "mx25l3205d",  INFO(0xc22016, 0, 64 * 1024,  64, 0) },
+	{ "mx25l3235d",  INFO(0xc25e16, 0, 64 * 1024,  64, SECT_4K) },
+	{ "mx25l3255e",  INFO(0xc29e16, 0, 64 * 1024,  64, SECT_4K) },
+	{ "mx25l6405d",  INFO(0xc22017, 0, 64 * 1024, 128, 0) },
+	{ "mx25l12805d", INFO(0xc22018, 0, 64 * 1024, 256, 0) },
+	{ "mx25l12855e", INFO(0xc22618, 0, 64 * 1024, 256, 0) },
+	{ "mx25l25635e", INFO(0xc22019, 0, 64 * 1024, 512, 0) },
+	{ "mx25l25655e", INFO(0xc22619, 0, 64 * 1024, 512, 0) },
+	{ "mx66l51235l", INFO(0xc2201a, 0, 64 * 1024, 1024, M25P80_QUAD_READ) },
+	{ "mx66l1g45g", INFO(0xc2201b, 0, 64 * 1024, 2048, 0) },
+
+	/* Micron */
+	{ "n25q064",  INFO(0x20ba17, 0, 64 * 1024, 128, 0) },
+	{ "n25q128a11",  INFO(0x20bb18, 0, 64 * 1024, 256, 0) },
+	{ "n25q128a13",  INFO(0x20ba18, 0, 64 * 1024, 256, 0) },
+	{ "n25q256a", INFO(0x20ba19, 0, 64 * 1024, 512, SECT_4K) },
+	{ "n25q512a",    INFO(0x20bb20, 0, 64 * 1024, 1024, SECT_4K) },
+
+	/* PMC */
+	{ "pm25lv512", INFO(0, 0, 32 * 1024, 2, SECT_4K_PMC) },
+	{ "pm25lv010", INFO(0, 0, 32 * 1024, 4, SECT_4K_PMC) },
+	{ "pm25lq032", INFO(0x7f9d46, 0, 64 * 1024,  64, SECT_4K) },
+
+	/* Spansion -- single (large) sector size only, at least
+	 * for the chips listed here (without boot sectors).
+	 */
+	{ "s25sl032p",  INFO(0x010215, 0x4d00,  64 * 1024,  64, 0) },
+	{ "s25sl064p",  INFO(0x010216, 0x4d00,  64 * 1024, 128, 0) },
+	{ "s25fl256s0", INFO(0x010219, 0x4d00, 256 * 1024, 128, 0) },
+	{ "s25fl256s1", INFO(0x010219, 0x4d01,  64 * 1024, 512, M25P80_QUAD_READ) },
+	{ "s25fl512s",  INFO(0x010220, 0x4d00, 256 * 1024, 256, M25P80_QUAD_READ) },
+	{ "s70fl01gs",  INFO(0x010221, 0x4d00, 256 * 1024, 256, 0) },
+	{ "s25sl12800", INFO(0x012018, 0x0300, 256 * 1024,  64, 0) },
+	{ "s25sl12801", INFO(0x012018, 0x0301,  64 * 1024, 256, 0) },
+	{ "s25fl129p0", INFO(0x012018, 0x4d00, 256 * 1024,  64, 0) },
+	{ "s25fl129p1", INFO(0x012018, 0x4d01,  64 * 1024, 256, 0) },
+	{ "s25sl004a",  INFO(0x010212,      0,  64 * 1024,   8, 0) },
+	{ "s25sl008a",  INFO(0x010213,      0,  64 * 1024,  16, 0) },
+	{ "s25sl016a",  INFO(0x010214,      0,  64 * 1024,  32, 0) },
+	{ "s25sl032a",  INFO(0x010215,      0,  64 * 1024,  64, 0) },
+	{ "s25sl064a",  INFO(0x010216,      0,  64 * 1024, 128, 0) },
+	{ "s25fl016k",  INFO(0xef4015,      0,  64 * 1024,  32, SECT_4K) },
+	{ "s25fl064k",  INFO(0xef4017,      0,  64 * 1024, 128, SECT_4K) },
+
+	/* SST -- large erase sizes are "overlays", "sectors" are 4K */
+	{ "sst25vf040b", INFO(0xbf258d, 0, 64 * 1024,  8, SECT_4K | SST_WRITE) },
+	{ "sst25vf080b", INFO(0xbf258e, 0, 64 * 1024, 16, SECT_4K | SST_WRITE) },
+	{ "sst25vf016b", INFO(0xbf2541, 0, 64 * 1024, 32, SECT_4K | SST_WRITE) },
+	{ "sst25vf032b", INFO(0xbf254a, 0, 64 * 1024, 64, SECT_4K | SST_WRITE) },
+	{ "sst25vf064c", INFO(0xbf254b, 0, 64 * 1024, 128, SECT_4K) },
+	{ "sst25wf512",  INFO(0xbf2501, 0, 64 * 1024,  1, SECT_4K | SST_WRITE) },
+	{ "sst25wf010",  INFO(0xbf2502, 0, 64 * 1024,  2, SECT_4K | SST_WRITE) },
+	{ "sst25wf020",  INFO(0xbf2503, 0, 64 * 1024,  4, SECT_4K | SST_WRITE) },
+	{ "sst25wf040",  INFO(0xbf2504, 0, 64 * 1024,  8, SECT_4K | SST_WRITE) },
+
+	/* ST Microelectronics -- newer production may have feature updates */
+	{ "m25p05",  INFO(0x202010,  0,  32 * 1024,   2, 0) },
+	{ "m25p10",  INFO(0x202011,  0,  32 * 1024,   4, 0) },
+	{ "m25p20",  INFO(0x202012,  0,  64 * 1024,   4, 0) },
+	{ "m25p40",  INFO(0x202013,  0,  64 * 1024,   8, 0) },
+	{ "m25p80",  INFO(0x202014,  0,  64 * 1024,  16, 0) },
+	{ "m25p16",  INFO(0x202015,  0,  64 * 1024,  32, 0) },
+	{ "m25p32",  INFO(0x202016,  0,  64 * 1024,  64, 0) },
+	{ "m25p64",  INFO(0x202017,  0,  64 * 1024, 128, 0) },
+	{ "m25p128", INFO(0x202018,  0, 256 * 1024,  64, 0) },
+	{ "n25q032", INFO(0x20ba16,  0,  64 * 1024,  64, 0) },
+
+	{ "m25p05-nonjedec",  INFO(0, 0,  32 * 1024,   2, 0) },
+	{ "m25p10-nonjedec",  INFO(0, 0,  32 * 1024,   4, 0) },
+	{ "m25p20-nonjedec",  INFO(0, 0,  64 * 1024,   4, 0) },
+	{ "m25p40-nonjedec",  INFO(0, 0,  64 * 1024,   8, 0) },
+	{ "m25p80-nonjedec",  INFO(0, 0,  64 * 1024,  16, 0) },
+	{ "m25p16-nonjedec",  INFO(0, 0,  64 * 1024,  32, 0) },
+	{ "m25p32-nonjedec",  INFO(0, 0,  64 * 1024,  64, 0) },
+	{ "m25p64-nonjedec",  INFO(0, 0,  64 * 1024, 128, 0) },
+	{ "m25p128-nonjedec", INFO(0, 0, 256 * 1024,  64, 0) },
+
+	{ "m45pe10", INFO(0x204011,  0, 64 * 1024,    2, 0) },
+	{ "m45pe80", INFO(0x204014,  0, 64 * 1024,   16, 0) },
+	{ "m45pe16", INFO(0x204015,  0, 64 * 1024,   32, 0) },
+
+	{ "m25pe20", INFO(0x208012,  0, 64 * 1024,  4,       0) },
+	{ "m25pe80", INFO(0x208014,  0, 64 * 1024, 16,       0) },
+	{ "m25pe16", INFO(0x208015,  0, 64 * 1024, 32, SECT_4K) },
+
+	{ "m25px16",    INFO(0x207115,  0, 64 * 1024, 32, SECT_4K) },
+	{ "m25px32",    INFO(0x207116,  0, 64 * 1024, 64, SECT_4K) },
+	{ "m25px32-s0", INFO(0x207316,  0, 64 * 1024, 64, SECT_4K) },
+	{ "m25px32-s1", INFO(0x206316,  0, 64 * 1024, 64, SECT_4K) },
+	{ "m25px64",    INFO(0x207117,  0, 64 * 1024, 128, 0) },
+
+	/* Winbond -- w25x "blocks" are 64K, "sectors" are 4KiB */
+	{ "w25x05", INFO(0xef3010, 0, 64 * 1024,  1,  SECT_4K) },
+	{ "w25x10", INFO(0xef3011, 0, 64 * 1024,  2,  SECT_4K) },
+	{ "w25x20", INFO(0xef3012, 0, 64 * 1024,  4,  SECT_4K) },
+	{ "w25x40", INFO(0xef3013, 0, 64 * 1024,  8,  SECT_4K) },
+	{ "w25x80", INFO(0xef3014, 0, 64 * 1024,  16, SECT_4K) },
+	{ "w25x16", INFO(0xef3015, 0, 64 * 1024,  32, SECT_4K) },
+	{ "w25x32", INFO(0xef3016, 0, 64 * 1024,  64, SECT_4K) },
+	{ "w25q16", INFO(0xef4015, 0, 64 * 1024,  32, SECT_4K) },
+	{ "w25q32", INFO(0xef4016, 0, 64 * 1024,  64, SECT_4K) },
+	{ "w25q32dw", INFO(0xef6016, 0, 64 * 1024,  64, SECT_4K) },
+	{ "w25x64", INFO(0xef3017, 0, 64 * 1024, 128, SECT_4K) },
+	{ "w25q64", INFO(0xef4017, 0, 64 * 1024, 128, SECT_4K) },
+	{ "w25q128", INFO(0xef4018, 0, 64 * 1024, 256, SECT_4K) },
+	{ "w25q80", INFO(0xef5014, 0, 64 * 1024,  16, SECT_4K) },
+	{ "w25q80bl", INFO(0xef4014, 0, 64 * 1024,  16, SECT_4K) },
+	{ "w25q128", INFO(0xef4018, 0, 64 * 1024, 256, SECT_4K) },
+	{ "w25q256", INFO(0xef4019, 0, 64 * 1024, 512, SECT_4K) },
+
+	/* Catalyst / On Semiconductor -- non-JEDEC */
+	{ "cat25c11", CAT25_INFO(  16, 8, 16, 1, M25P_NO_ERASE | M25P_NO_FR) },
+	{ "cat25c03", CAT25_INFO(  32, 8, 16, 2, M25P_NO_ERASE | M25P_NO_FR) },
+	{ "cat25c09", CAT25_INFO( 128, 8, 32, 2, M25P_NO_ERASE | M25P_NO_FR) },
+	{ "cat25c17", CAT25_INFO( 256, 8, 32, 2, M25P_NO_ERASE | M25P_NO_FR) },
+	{ "cat25128", CAT25_INFO(2048, 8, 64, 2, M25P_NO_ERASE | M25P_NO_FR) },
+	{ },
+};
+MODULE_DEVICE_TABLE(spi, m25p_ids);
+
+static const struct spi_device_id *jedec_probe(struct spi_device *spi)
+{
+	int			tmp;
+	u8			code = OPCODE_RDID;
+	u8			id[5];
+	u32			jedec, default_jedec;
+	u16                     ext_jedec;
+	struct flash_info	*info;
+
+	/* JEDEC also defines an optional "extended device information"
+	 * string for after vendor-specific data, after the three bytes
+	 * we use here.  Supporting some chips might require using it.
+	 */
+	tmp = spi_write_then_read(spi, &code, 1, id, 5);
+	if (tmp < 0) {
+		pr_debug("\n>>>>>%s: error %d reading JEDEC ID<<<<<\n\n",
+				dev_name(&spi->dev), tmp);
+		return ERR_PTR(tmp);
+	}
+	jedec = id[0];
+	jedec = jedec << 8;
+	jedec |= id[1];
+	jedec = jedec << 8;
+	jedec |= id[2];
+
+	ext_jedec = id[3] << 8 | id[4];
+	printk("JEDEC id %06X\n", jedec);
+	for (tmp = 0; tmp < ARRAY_SIZE(m25p_ids) - 1; tmp++) {
+		info = (void *)m25p_ids[tmp].driver_data;
+		if (info->jedec_id == jedec) {
+			if (info->ext_id != 0 && info->ext_id != ext_jedec)
+				continue;
+			return &m25p_ids[tmp];
+		}
+	}
+	dev_err(&spi->dev, "\n>>>>>unrecognized JEDEC id %06x<<<<<\n\n", jedec);
+	// if unrecognized, use mx25l3205d as default
+	printk("use JEDEC id 0xc22016 (mx25l3205d) instead\n");
+	default_jedec = 0xc22016;
+	for (tmp = 0; tmp < ARRAY_SIZE(m25p_ids) - 1; tmp++) {
+		info = (void *)m25p_ids[tmp].driver_data;
+		if (info->jedec_id == default_jedec)
+			return &m25p_ids[tmp];
+	}
+	return ERR_PTR(-ENODEV);
 }

+
 /*
  * board specific setup should have ensured the SPI clock used here
  * matches what the READ command supports, at least until this driver
@@ -191,61 +1515,301 @@
  */
 static int m25p_probe(struct spi_device *spi)
 {
-	struct mtd_part_parser_data	ppdata;
+	const struct spi_device_id	*id = spi_get_device_id(spi);
	struct flash_platform_data	*data;
-	struct m25p *flash;
-	struct spi_nor *nor;
-	enum read_mode mode = SPI_NOR_NORMAL;
-	char *flash_name = NULL;
+	struct m25p			*flash;
+	struct flash_info		*info;
+	unsigned			i;
+	struct mtd_part_parser_data	ppdata;
+#ifdef CONFIG_OF
+	struct device_node *np = spi->dev.of_node;
+#endif
	int ret;

+	/* Platform data helps sort out which chip type we have, as
+	 * well as how this board partitions it.  If we don't have
+	 * a chip ID, try the JEDEC id commands; they'll work for most
+	 * newer chips, even if we don't recognize the particular chip.
+	 */
	data = dev_get_platdata(&spi->dev);
+	if (data && data->type) {
+		const struct spi_device_id *plat_id;
+
+		for (i = 0; i < ARRAY_SIZE(m25p_ids) - 1; i++) {
+			plat_id = &m25p_ids[i];
+			if (strcmp(data->type, plat_id->name))
+				continue;
+			break;
+		}
+
+		if (i < ARRAY_SIZE(m25p_ids) - 1)
+			id = plat_id;
+		else
+			dev_warn(&spi->dev, "unrecognized id %s\n", data->type);
+	}
+
+	info = (void *)id->driver_data;
+
+	if (info->jedec_id) {
+		const struct spi_device_id *jid;
+
+		jid = jedec_probe(spi);
+		if (IS_ERR(jid)) {
+			return PTR_ERR(jid);
+		} else if (jid != id) {
+			/*
+			 * JEDEC knows better, so overwrite platform ID. We
+			 * can't trust partitions any longer, but we'll let
+			 * mtd apply them anyway, since some partitions may be
+			 * marked read-only, and we don't want to lose that
+			 * information, even if it's not 100% accurate.
+			 */
+			dev_warn(&spi->dev, "found %s, expected %s\n",
+				 jid->name, id->name);
+			id = jid;
+			info = (void *)jid->driver_data;
+		}
+	}

	flash = devm_kzalloc(&spi->dev, sizeof(*flash), GFP_KERNEL);
	if (!flash)
		return -ENOMEM;

-	nor = &flash->spi_nor;
-
-	/* install the hooks */
-	nor->read = m25p80_read;
-	nor->write = m25p80_write;
-	nor->erase = m25p80_erase;
-	nor->write_reg = m25p80_write_reg;
-	nor->read_reg = m25p80_read_reg;
-
-	nor->dev = &spi->dev;
-	nor->mtd = &flash->mtd;
-	nor->priv = flash;
+	flash->command = devm_kzalloc(&spi->dev, MAX_CMD_SIZE, GFP_KERNEL);
+	if (!flash->command)
+		return -ENOMEM;

-	spi_set_drvdata(spi, flash);
-	flash->mtd.priv = nor;
	flash->spi = spi;
+	mutex_init(&flash->lock);
+	spi_set_drvdata(spi, flash);

-	if (spi->mode & SPI_RX_QUAD)
-		mode = SPI_NOR_QUAD;
-	else if (spi->mode & SPI_RX_DUAL)
-		mode = SPI_NOR_DUAL;
+	/*
+	 * Atmel, SST and Intel/Numonyx serial flash tend to power
+	 * up with the software protection bits set
+	 */
+
+	if (JEDEC_MFR(info->jedec_id) == CFI_MFR_ATMEL ||
+	    JEDEC_MFR(info->jedec_id) == CFI_MFR_INTEL ||
+	    JEDEC_MFR(info->jedec_id) == CFI_MFR_SST) {
+		write_enable(flash);
+		write_sr(flash, 0);
+	}

	if (data && data->name)
		flash->mtd.name = data->name;
+	else
+		flash->mtd.name = dev_name(&spi->dev);

-	/* For some (historical?) reason many platforms provide two different
-	 * names in flash_platform_data: "name" and "type". Quite often name is
-	 * set to "m25p80" and then "type" provides a real chip name.
-	 * If that's the case, respect "type" and ignore a "name".
-	 */
-	if (data && data->type)
-		flash_name = data->type;
+	flash->mtd.type = MTD_NORFLASH;
+	flash->mtd.writesize = 1;
+	flash->mtd.flags = MTD_CAP_NORFLASH;
+	flash->mtd.size = info->sector_size * info->n_sectors;
+	flash->mtd._erase = m25p80_erase;
+	flash->mtd._read = m25p80_read;
+
+	/* flash protection support for STmicro chips */
+	if (JEDEC_MFR(info->jedec_id) == CFI_MFR_ST) {
+		flash->mtd._lock = m25p80_lock;
+		flash->mtd._unlock = m25p80_unlock;
+	}
+
+	/* sst flash chips use AAI word program */
+	if (info->flags & SST_WRITE)
+		flash->mtd._write = sst_write;
	else
-		flash_name = spi->modalias;
+		flash->mtd._write = m25p80_write;

-	ret = spi_nor_scan(nor, flash_name, mode);
-	if (ret)
-		return ret;
+	/* prefer "small sector" erase if possible */
+	if (info->flags & SECT_4K) {
+		flash->erase_opcode = OPCODE_BE_4K;
+		flash->mtd.erasesize = 4096;
+	} else if (info->flags & SECT_4K_PMC) {
+		flash->erase_opcode = OPCODE_BE_4K_PMC;
+		flash->mtd.erasesize = 4096;
+	} else {
+		flash->erase_opcode = OPCODE_SE;
+		flash->mtd.erasesize = info->sector_size;
+	}
+
+	if (info->flags & M25P_NO_ERASE)
+		flash->mtd.flags |= MTD_NO_ERASE;

	ppdata.of_node = spi->dev.of_node;
+	flash->mtd.dev.parent = &spi->dev;
+	flash->page_size = info->page_size;
+	flash->mtd.writebufsize = flash->page_size;
+
+	flash->fast_read = false;
+	flash->quad = false;
+	flash->dual = false;
+	flash->auto_mode = false;
+#ifdef CONFIG_OF
+	if (np) {
+		/* If we were instantiated by DT, use it */
+		if (of_property_read_bool(np, "m25p,fast-read"))
+			flash->flash_read = M25P80_FAST;
+		else
+			flash->flash_read = M25P80_NORMAL;
+	} else {
+		/* If we weren't instantiated by DT, default to fast-read */
+		flash->flash_read = M25P80_FAST;
+	}
+#endif
+
+#ifdef CONFIG_M25PXX_USE_MULTI_CHANNEL
+#ifdef CONFIG_M25PXX_USE_QUAD
+	flash->quad = true;
+#endif
+#ifdef CONFIG_M25PXX_USE_DUAL
+	flash->dual = true;
+#endif
+#else
+#ifdef CONFIG_M25PXX_USE_FAST_READ
+	flash->fast_read = true;
+#endif
+#endif
+#ifdef CONFIG_M25PXX_USE_AUTO_MODE
+	flash->auto_mode = true;
+#endif
+	/* Some devices cannot do fast-read, no matter what DT tells us */
+	if (info->flags & M25P_NO_FR)
+		flash->flash_read = M25P80_NORMAL;
+
+	/* Quad-read mode takes precedence over fast/normal */
+	if (spi->mode & SPI_RX_QUAD && info->flags & M25P80_QUAD_READ) {
+		ret = set_quad_mode(flash, info->jedec_id);
+		if (ret) {
+			dev_err(&flash->spi->dev, "quad mode not supported\n");
+			return ret;
+		}
+		flash->flash_read = M25P80_QUAD;
+	}

+	/* flash-specific code  */
+	ret = get_flash_device(flash, info->jedec_id);
+	if (ret) {
+		dev_err(&flash->spi->dev, "no flash device supported\n");
+		return ret;
+	}
+
+	if (flash->auto_mode) {
+		flash->flash_read = M25P80_AUTO;
+		flash->flash_write = M25P80_AUTO_WRITE;
+	} else if (flash->quad) {
+		flash->flash_read = M25P80_QUAD;
+		flash->flash_write = M25P80_QUAD_WRITE;
+	} else if (flash->dual) {
+		flash->flash_read = M25P80_DUAL;
+		flash->flash_write = M25P80_DUAL_WRITE;
+	} else if (flash->fast_read) {
+		flash->flash_read = M25P80_FAST;
+		flash->flash_write = M25P80_NORMAL_WRITE;
+	} else {
+		flash->flash_read = M25P80_NORMAL;
+		flash->flash_write = M25P80_NORMAL_WRITE;
+	}
+
+	/* Default commands */
+	switch (flash->flash_read) {
+	case M25P80_QUAD:
+		flash->read_opcode = flash->dev_info.cmd->readx4_io;
+		flash->dummy = flash->dev_info.dummy->rd_quad_dummy;
+		flash->read_type = flash->dev_info.type->rd_quad_type;
+		break;
+	case M25P80_DUAL:
+		flash->read_opcode = flash->dev_info.cmd->readx2_io;
+		flash->dummy = flash->dev_info.dummy->rd_dual_dummy;
+		flash->read_type = flash->dev_info.type->rd_dual_type;
+		break;
+	case M25P80_FAST:
+		flash->read_opcode = OPCODE_FAST_READ;
+		flash->dummy = flash->dev_info.dummy->fast_rd_dummy;
+		break;
+	case M25P80_NORMAL:
+		flash->read_opcode = OPCODE_NORM_READ;
+		break;
+	case M25P80_AUTO:
+		flash->read_opcode = OPCODE_AUTO;
+		break;
+	default:
+		dev_err(&flash->spi->dev, "No Read opcode defined\n");
+		return -EINVAL;
+	}
+
+	switch (flash->flash_write) {
+	case M25P80_QUAD_WRITE:
+		flash->program_opcode = flash->dev_info.cmd->ppx4_ii;
+		flash->write_type = flash->dev_info.type->wr_quad_type;
+		break;
+	case M25P80_DUAL_WRITE:
+		flash->program_opcode = flash->dev_info.cmd->ppx2_ii;
+		flash->write_type = flash->dev_info.type->wr_dual_type;
+		break;
+	case M25P80_NORMAL_WRITE:
+		flash->program_opcode = OPCODE_PP;
+		break;
+	case M25P80_AUTO_WRITE:
+		flash->program_opcode = OPCODE_AUTO;
+		break;
+	default:
+		dev_err(&flash->spi->dev, "No write opcode defined\n");
+		return -EINVAL;
+	}
+
+	if (info->addr_width)
+		flash->addr_width = info->addr_width;
+	else if (flash->mtd.size > 0x1000000) {
+		/* enable 4-byte addressing if the device exceeds 16MiB */
+			flash->addr_width = 4;
+		if (JEDEC_MFR(info->jedec_id) == CFI_MFR_AMD) {
+			/* Dedicated 4-byte command set */
+			switch (flash->flash_read) {
+			case M25P80_QUAD:
+				flash->read_opcode = OPCODE_QUAD_READ_4B;
+				break;
+			case M25P80_FAST:
+				flash->read_opcode = OPCODE_FAST_READ_4B;
+				break;
+			case M25P80_NORMAL:
+			case M25P80_DUAL:
+			case M25P80_AUTO:
+				flash->read_opcode = OPCODE_NORM_READ_4B;
+				break;
+			}
+			flash->program_opcode = OPCODE_PP_4B;
+			/* No small sector erase for 4-byte command set */
+			flash->erase_opcode = OPCODE_SE_4B;
+			flash->mtd.erasesize = info->sector_size;
+		} else
+			set_4byte(flash, info->jedec_id, 1);
+	} else {
+			flash->addr_width = 3;
+	}
+
+	dev_info(&spi->dev, "%s (%lld Kbytes)\n", id->name,
+			(long long)flash->mtd.size >> 10);
+
+	pr_debug("mtd .name = %s, .size = 0x%llx (%lldMiB) "
+			".erasesize = 0x%.8x (%uKiB) .numeraseregions = %d\n",
+		flash->mtd.name,
+		(long long)flash->mtd.size, (long long)(flash->mtd.size >> 20),
+		flash->mtd.erasesize, flash->mtd.erasesize / 1024,
+		flash->mtd.numeraseregions);
+
+	if (flash->mtd.numeraseregions)
+		for (i = 0; i < flash->mtd.numeraseregions; i++)
+			pr_debug("mtd.eraseregions[%d] = { .offset = 0x%llx, "
+				".erasesize = 0x%.8x (%uKiB), "
+				".numblocks = %d }\n",
+				i, (long long)flash->mtd.eraseregions[i].offset,
+				flash->mtd.eraseregions[i].erasesize,
+				flash->mtd.eraseregions[i].erasesize / 1024,
+				flash->mtd.eraseregions[i].numblocks);
+
+
+	/* partitions should match sector boundaries; and it may be good to
+	 * use readonly partitions for writeprotected sectors (BP2..BP0).
+	 */
	return mtd_device_parse_register(&flash->mtd, NULL, &ppdata,
			data ? data->parts : NULL,
			data ? data->nr_parts : 0);
@@ -261,54 +1825,6 @@
 }


-/*
- * XXX This needs to be kept in sync with spi_nor_ids.  We can't share
- * it with spi-nor, because if this is built as a module then modpost
- * won't be able to read it and add appropriate aliases.
- */
-static const struct spi_device_id m25p_ids[] = {
-	{"at25fs010"},	{"at25fs040"},	{"at25df041a"},	{"at25df321a"},
-	{"at25df641"},	{"at26f004"},	{"at26df081a"},	{"at26df161a"},
-	{"at26df321"},	{"at45db081d"},
-	{"en25f32"},	{"en25p32"},	{"en25q32b"},	{"en25p64"},
-	{"en25q64"},	{"en25qh128"},	{"en25qh256"},
-	{"f25l32pa"},
-	{"mr25h256"},	{"mr25h10"},
-	{"gd25q32"},	{"gd25q64"},
-	{"160s33b"},	{"320s33b"},	{"640s33b"},
-	{"mx25l2005a"},	{"mx25l4005a"},	{"mx25l8005"},	{"mx25l1606e"},
-	{"mx25l3205d"},	{"mx25l3255e"},	{"mx25l6405d"},	{"mx25l12805d"},
-	{"mx25l12855e"},{"mx25l25635e"},{"mx25l25655e"},{"mx66l51235l"},
-	{"mx66l1g55g"},
-	{"n25q064"},	{"n25q128a11"},	{"n25q128a13"},	{"n25q256a"},
-	{"n25q512a"},	{"n25q512ax3"},	{"n25q00"},
-	{"pm25lv512"},	{"pm25lv010"},	{"pm25lq032"},
-	{"s25sl032p"},	{"s25sl064p"},	{"s25fl256s0"},	{"s25fl256s1"},
-	{"s25fl512s"},	{"s70fl01gs"},	{"s25sl12800"},	{"s25sl12801"},
-	{"s25fl129p0"},	{"s25fl129p1"},	{"s25sl004a"},	{"s25sl008a"},
-	{"s25sl016a"},	{"s25sl032a"},	{"s25sl064a"},	{"s25fl008k"},
-	{"s25fl016k"},	{"s25fl064k"},
-	{"sst25vf040b"},{"sst25vf080b"},{"sst25vf016b"},{"sst25vf032b"},
-	{"sst25vf064c"},{"sst25wf512"},	{"sst25wf010"},	{"sst25wf020"},
-	{"sst25wf040"},
-	{"m25p05"},	{"m25p10"},	{"m25p20"},	{"m25p40"},
-	{"m25p80"},	{"m25p16"},	{"m25p32"},	{"m25p64"},
-	{"m25p128"},	{"n25q032"},
-	{"m25p05-nonjedec"},	{"m25p10-nonjedec"},	{"m25p20-nonjedec"},
-	{"m25p40-nonjedec"},	{"m25p80-nonjedec"},	{"m25p16-nonjedec"},
-	{"m25p32-nonjedec"},	{"m25p64-nonjedec"},	{"m25p128-nonjedec"},
-	{"m45pe10"},	{"m45pe80"},	{"m45pe16"},
-	{"m25pe20"},	{"m25pe80"},	{"m25pe16"},
-	{"m25px16"},	{"m25px32"},	{"m25px32-s0"},	{"m25px32-s1"},
-	{"m25px64"},	{"m25px80"},
-	{"w25x10"},	{"w25x20"},	{"w25x40"},	{"w25x80"},
-	{"w25x16"},	{"w25x32"},	{"w25q32"},	{"w25q32dw"},
-	{"w25x64"},	{"w25q64"},	{"w25q80"},	{"w25q80bl"},
-	{"w25q128"},	{"w25q256"},	{"cat25c11"},
-	{"cat25c03"},	{"cat25c09"},	{"cat25c17"},	{"cat25128"},
-	{ },
-};
-MODULE_DEVICE_TABLE(spi, m25p_ids);
 
 static const struct of_device_id m25p_of_table[] = {
 	/*
