diff '--color=auto' -aur a/block.c b/block.c
--- a/block.c	2020-05-11 19:41:59.000000000 -0400
+++ b/block.c	2022-09-06 20:55:10.657693763 -0400
@@ -697,18 +697,6 @@
 	return 0;
 }
 
-static void mkdir_p(char *dir)
-{
-	char *l = strrchr(dir, '/');
-
-	if (l) {
-		*l = '\0';
-		mkdir_p(dir);
-		*l = '/';
-		mkdir(dir, 0755);
-	}
-}
-
 static void check_filesystem(struct probe_info *pr)
 {
 	pid_t pid;
@@ -1085,7 +1073,7 @@
 	if (check_fs)
 		check_filesystem(pr);
 
-	mkdir_p(target);
+	mkdir_p(target, 0755);
 	if (!lstat(target, &st) && S_ISLNK(st.st_mode))
 		unlink(target);
 
@@ -1396,7 +1384,7 @@
 
 	snprintf(tag, sizeof(tag), "%s/etc", path);
 	if (stat(tag, &s))
-		mkdir_p(tag);
+		mkdir_p(tag, 0755);
 
 	snprintf(tag, sizeof(tag), "%s/etc/.extroot-uuid", path);
 	if (stat(tag, &s)) {
@@ -1486,7 +1474,7 @@
 
 		if (m->overlay)
 			path = overlay;
-		mkdir_p(path);
+		mkdir_p(path, 0755);
 
 		if (check_fs)
 			check_filesystem(pr);
@@ -1556,7 +1544,7 @@
 			 * Mount MTD part and try extroot (using
 			 * /etc/config/fstab from that partition)
 			 */
-			mkdir_p(cfg);
+			mkdir_p(cfg, 0755);
 			if (!mount(blkdev_path, cfg, "jffs2", MS_NOATIME, NULL)) {
 				err = mount_extroot(cfg);
 				umount2(cfg, MNT_DETACH);
@@ -1578,7 +1566,7 @@
 		char cfg[] = "/tmp/ubifs_cfg";
 
 		/* Mount volume and try extroot (using fstab from that vol) */
-		mkdir_p(cfg);
+		mkdir_p(cfg, 0755);
 		if (!mount(blkdev_path, cfg, "ubifs", MS_NOATIME, NULL)) {
 			err = mount_extroot(cfg);
 			umount2(cfg, MNT_DETACH);
